src/ast.js

```diff
@@ -2679,6 +2679,16 @@ function parsenumber (c, s, lineno) {
     var tmp;
     var end = s.charAt(s.length - 1);
 
+    if (/_[eE]|[eE]_|\._|j_/.test(s)) {
+        throw new Sk.builtin.SyntaxError("invalid syntax", c.c_filename, lineno);
+    }
+
+    if (/_\.|[+-]_|^0_\D|_j/.test(s)) {
+        throw new Sk.builtin.SyntaxError("invalid decimal literal", c.c_filename, lineno);
+    }
+    
+    s = s.replace(/_(?=[^_])/g, "");
+    
```

Comment:
worth hositing the regex in this pr - i.e. defining them outside the scope of the function so that they are only compiled once. This can improve performance.